mongo/src/main/java/com/redhat/lightblue/mongo/crud/MongoSafeUpdateProtocol.java

```diff
@@ -0,0 +1,349 @@
+                index=0;
+                for(Object id:idsInBatch) {
+                    if(!results.containsKey(index)) { // No other errors for this id
+                        if(!successfulIds.contains(id)) {
+                            // concurrency errors for this id
+                            results.put(index,Error.get("update",MongoCrudConstants.ERR_CONCURRENT_UPDATE,id.toString()));
+                        }
+                    }
+                    index++;
+                }
+            } finally {
+                if(cursor!=null)
+                    cursor.close();
+            }
+        }
+    }
+
+    /**
+     * Writes a replace query, which is:
+     * <pre>
+     *  {_id:<docId>, "@mongoHidden.docver.0":ObjectId(<originalDocver>)}
+     * </pre>
+     */
+    private DBObject writeReplaceQuery(DBObject doc) {
+        DBObject hidden=getHidden(doc,false);
+        ObjectId top=null;
+        if(hidden!=null) {
+            List<ObjectId> list=(List<ObjectId>)hidden.get(DOCVER);
+            if(list!=null&&!list.isEmpty())
+                top=list.get(0);
+        }
+        BasicDBObject query=new BasicDBObject("_id",doc.get("_id"));
+        query.append(DOCVER_FLD0,top);
+        return query;
+    }
+    
+
+    /**
+     * Returns the @mongoHidden at the root level of the document. Adds one if necessary.
+     */
+    public static DBObject getHidden(DBObject doc,boolean addIfNotFound) {
+        DBObject hidden=(DBObject)doc.get(Translator.HIDDEN_SUB_PATH.toString());
+        if(hidden==null&&addIfNotFound) {
+            doc.put(Translator.HIDDEN_SUB_PATH.toString(),hidden=new BasicDBObject());            
+        }
+        return hidden;
+    }
+
+    public static void overwriteDocVer(DBObject doc,ObjectId docver) {
+        getHidden(doc,true).removeField(DOCVER);
+        setDocVer(doc,docver);
+    }
+    
+    public static void setDocVer(DBObject doc,ObjectId docver) {
+        DBObject hidden=getHidden(doc,true);
+        List<ObjectId> list=(List<ObjectId>)hidden.get(DOCVER);
+        if(list==null) {
+            list=new ArrayList<ObjectId>();
+        }
+        list.add(0,docver);
+        hidden.put(DOCVER,list);
+    }
+
+    /**
+     * Removes old docvers from a document
+     */
+    private void cleanupOldDocVer(DBObject doc) {
+        DBObject hidden=getHidden(doc,false);
+        if(hidden!=null) {
+            List<ObjectId> list=(List<ObjectId>)hidden.get(DOCVER);
+            if(list!=null) {
+                List<ObjectId> copy=new ArrayList<>(list.size());
+                long now=docVer.getDate().getTime();
+                for(ObjectId id:list) {
+                    if(!id.equals(docVer)) {
+                        Date d=id.getDate();
+                        if(now-d.getTime()<TOO_OLD) {
+                            copy.add(id);
+                        }
+                    } else {
+                        copy.add(id);
+                    }
+                }
```

Comment:
Hrm... concurrency is always a bit mind bending :).

The scenario that could be a problem (but perhaps it is too rare depending on timeout?) I think is this:

Given timeout is 1:00 (1 minute), and there are [at least] three overlapping updates throughout this timeframe.

0. Two updates, A and B find their docs at ~0:00
1. A and B start commits at 0:01 with docvers from ~0:00
2. They finish. B is successful, returns. A misses some updates, will check for concurrent mods.
3. Update C starts at 1:00, removes some docvers from update A which are too old, and commits.
4. A finds concurrent mods at 1:01 (big delay from step 2... would this happen?), gets less back than expected because of update C removing old doc versions. It knows it updated N documents, but finds >N concurrent mods.

What do you think? Maybe to some extent we can't be perfect, and if timeout is long enough, it would be crazy to actually happen. So maybe imperfect is okay. (Assuming I don't have something wrong). Even my suggestion (keeping some old ones if array size not too big) is still imperfect (just imagine there are many successful updates in between 2 and 4 ... array size could be met eventually).